File: 05-AI与机器学习/01-深度学习/FlashAttention在线Softmax/go/flash.go
// Package flash —— Dao-AILab/flash-attention 的 flash_attn_triton.py 中 _fwd_kernel 的 Go 转写。
// 关键：源码维护的是 m_i（running max）与 lse_i（running log-sum-exp），不是常见的 (m, l) 组合。
package main

import "math"

// NegInf 负无穷（对应 Triton 里的 float("-inf")）。
var NegInf = math.Inf(-1)

// SoftmaxScaleFor 对应 softmax_scale = softmax_scale or 1.0 / math.sqrt(d)。
func SoftmaxScaleFor(d int, scale float64) float64 {
	if scale != 0 {
		return scale
	}
	return 1.0 / math.Sqrt(float64(d))
}

// SeqlenQRounded 对应 math.ceil(seqlen_q/128)*128。
func SeqlenQRounded(seqlenQ int) int {
	return int(math.Ceil(float64(seqlenQ)/128)) * 128
}

// Dot 内积。
func Dot(a, b []float64) float64 {
	s := 0.0
	for i := range a {
		s += a[i] * b[i]
	}
	return s
}

// NaiveAttention 朴素实现：整张 QK^T 落地（对照组）。
func NaiveAttention(q, k, v [][]float64, causal bool, scale float64, bias []float64) [][]float64 {
	sq, sk, d := len(q), len(k), len(v[0])
	out := make([][]float64, sq)
	for i := 0; i < sq; i++ {
		scores := make([]float64, sk)
		for j := 0; j < sk; j++ {
			s := Dot(q[i], k[j]) * scale
			if bias != nil {
				s += bias[j]
			}
			if causal && j > i {
				s = NegInf
			}
			scores[j] = s
		}
		mx := scores[0]
		for _, s := range scores {
			if s > mx {
				mx = s
			}
		}
		exps := make([]float64, sk)
		tot := 0.0
		for j, s := range scores {
			exps[j] = math.Exp(s - mx)
			tot += exps[j]
		}
		row := make([]float64, d)
		for j, e := range exps {
			w := e / tot
			for t := 0; t < d; t++ {
				row[t] += w * v[j][t]
			}
		}
		out[i] = row
	}
	return out
}

// FlashAttention 逐块在线 softmax；返回 (out, lse, scale)。
func FlashAttention(q, k, v [][]float64, causal bool, softmaxScale float64, bias []float64,
	blockM, blockN int, useLseForMax bool) ([][]float64, []float64, float64) {
	sq, sk, d := len(q), len(k), len(q[0])
	if d > 128 {
		panic("FlashAttention only support head dimensions up to 128")
	}
	scale := SoftmaxScaleFor(d, softmaxScale)
	out := make([][]float64, 0, sq)
	lse := make([]float64, 0, sq)
	for startM := 0; startM < sq; startM += blockM {
		end := startM + blockM
		if end > sq {
			end = sq
		}
		n := end - startM
		mI := make([]float64, n)
		lseI := make([]float64, n)
		acc := make([][]float64, n)
		for r := range mI {
			mI[r] = NegInf
			lseI[r] = NegInf
			acc[r] = make([]float64, d)
		}
		endN := sk
		if causal {
			if (startM+1)*blockM < sk {
				endN = (startM + 1) * blockM
			} else {
				endN = sk
			}
		}
		for startN := 0; startN < endN; startN += blockN {
			cn := startN + blockN
			if cn > sk {
				cn = sk
			}
			cols := make([]int, 0, cn-startN)
			for j := startN; j < cn; j++ {
				cols = append(cols, j)
			}
			for r := 0; r < n; r++ {
				i := startM + r
				rowMax := NegInf
				qk := make([]float64, len(cols))
				for ci, j := range cols {
					s := Dot(q[i], k[j]) * scale
					if bias != nil {
						s += bias[j]
					}
					if causal && j > i {
						s = NegInf
					}
					qk[ci] = s
					if s > rowMax {
						rowMax = s
					}
				}
				base := mI[r]
				if useLseForMax {
					base = lseI[r]
				}
				mIJ := rowMax
				if base > mIJ {
					mIJ = base
				}
				p := make([]float64, len(cols))
				lIJ := 0.0
				if math.IsInf(mIJ, -1) {
					for ci := range p {
						p[ci] = 0.0
					}
				} else {
					for ci, s := range qk {
						if math.IsInf(s, -1) {
							p[ci] = 0.0
							continue
						}
						p[ci] = math.Exp(s - mIJ)
						lIJ += p[ci]
					}
				}
				var accScale float64
				if math.IsInf(mIJ, -1) {
					accScale = 0.0
				} else {
					accScale = math.Exp(mI[r] - mIJ)
				}
				for t := 0; t < d; t++ {
					acc[r][t] *= accScale
				}
				for ci, e := range p {
					if e == 0.0 {
						continue
					}
					vj := v[cols[ci]]
					for t := 0; t < d; t++ {
						acc[r][t] += e * vj[t]
					}
				}
				newL := lIJ
				if !math.IsInf(mIJ, -1) && !math.IsInf(lseI[r], -1) {
					newL += math.Exp(lseI[r] - mIJ)
				}
				mI[r] = mIJ
				if newL > 0.0 && !math.IsInf(mIJ, -1) {
					lseI[r] = mIJ + math.Log(newL)
				} else {
					lseI[r] = NegInf
				}
			}
		}
		for r := 0; r < n; r++ {
			var oScale float64
			if math.IsInf(lseI[r], -1) {
				oScale = 0.0
			} else {
				oScale = math.Exp(mI[r] - lseI[r])
			}
			row := make([]float64, d)
			for t := 0; t < d; t++ {
				row[t] = acc[r][t] * oScale
			}
			out = append(out, row)
			lse = append(lse, lseI[r])
		}
	}
	return out, lse, scale
}

File: 05-AI与机器学习/01-深度学习/FlashAttention在线Softmax/python/flash.py
"""FlashAttention 在线 softmax —— Dao-AILab/flash-attention 的 flash_attn/flash_attn_triton.py 逐行转写。

核心是 _fwd_kernel 的「log-sum-exp 版在线 softmax」：
    m_ij  = max(max(qk·scale), lse_i)
    p     = exp(qk·scale − m_ij)
    l_ij  = Σ p
    acc  *= exp(m_i − m_ij)          # m_i 是**上一轮**的 max
    acc  += p @ v
    m_i   = m_ij
    lse_i = m_ij + log(exp(lse_i − m_ij) + l_ij)
    out   = acc * exp(m_i − lse_i)

形状（源码是 (batch, seqlen_q, nheads, d)，本 demo 取单 batch 单头）：
    q : [S_q][D]      k, v : [S_k][D]
"""

import math

NEG_INF = float("-inf")


def softmax_scale_for(d, scale=None):
    """_flash_attn_forward：softmax_scale = softmax_scale or 1/sqrt(d)。"""
    return scale if scale else 1.0 / math.sqrt(d)


def seqlen_q_rounded(seqlen_q):
    """_flash_attn_forward：math.ceil(seqlen_q / 128) * 128（lse/tmp 缓冲的对齐长度）。"""
    return math.ceil(seqlen_q / 128) * 128


def validate(q, k, v, d):
    """_flash_attn_forward 开头的断言：d ≤ 128。"""
    if d > 128:
        raise AssertionError("FlashAttention only support head dimensions up to 128")
    if not (len(q[0]) == len(k[0]) == len(v[0]) == d):
        raise AssertionError("All tensors must have the same head dimension")


def dot(a, b):
    return sum(x * y for x, y in zip(a, b))


def naive_attention(q, k, v, causal=False, softmax_scale=None, bias=None):
    """朴素实现：整张 QK^T 落地 + 全局 softmax（作为对照组）。"""
    sq, sk, d = len(q), len(k), len(v[0])
    scale = softmax_scale_for(d, softmax_scale)
    out = []
    for i in range(sq):
        scores = []
        for j in range(sk):
            s = dot(q[i], k[j]) * scale
            if bias is not None:
                s += bias[j]
            if causal and j > i:
                s = NEG_INF
            scores.append(s)
        mx = max(scores)
        exps = [math.exp(s - mx) for s in scores]
        tot = 0.0
        for e in exps:
            tot += e
        row = [0.0] * d
        for j, e in enumerate(exps):
            w = e / tot
            for t in range(d):
                row[t] += w * v[j][t]
        out.append(row)
    return out


def flash_attention(q, k, v, causal=False, softmax_scale=None, bias=None,
                    block_m=128, block_n=128, use_lse_for_max=True, d=None):
    """_fwd_kernel 的逐块在线 softmax。返回 (out, lse, softmax_scale)。

    use_lse_for_max=False 时把 m_ij 的基准换成 m_i（源码用的是 lse_i）。
    """
    sq, sk = len(q), len(k)
    dd = d if d else len(q[0])
    validate(q, k, v, dd)
    scale = softmax_scale_for(dd, softmax_scale)
    out, lse = [], []
    for start_m in range(0, sq, block_m):
        rows = range(start_m, min(start_m + block_m, sq))
        m_i = [NEG_INF] * len(rows)
        lse_i = [NEG_INF] * len(rows)
        acc = [[0.0] * dd for _ in rows]
        end_n = sk if not causal else min((start_m + 1) * block_m, sk)
        blocks = 0
        for start_n in range(0, end_n, block_n):
            cols = list(range(start_n, min(start_n + block_n, sk)))
            blocks += 1
            for r, i in enumerate(rows):
                qk = []
                for j in cols:
                    s = dot(q[i], k[j]) * scale
                    if bias is not None:
                        s += bias[j]
                    if causal and j > i:
                        s = NEG_INF
                    qk.append(s)
                row_max = max(qk) if qk else NEG_INF
                base = lse_i[r] if use_lse_for_max else m_i[r]
                m_ij = max(row_max, base)
                if m_ij == NEG_INF:
                    # 全掩码且尚无历史（源码默认 BLOCK_M == BLOCK_N 时不会出现）
                    p = [0.0] * len(cols)
                    l_ij = 0.0
                else:
                    p = [math.exp(s - m_ij) if s != NEG_INF else 0.0 for s in qk]
                    l_ij = 0.0
                    for e in p:
                        l_ij += e
                acc_scale = math.exp(m_i[r] - m_ij) if m_ij != NEG_INF else 0.0
                for t in range(dd):
                    acc[r][t] *= acc_scale
                for jj, e in enumerate(p):
                    if e == 0.0:
                        continue
                    vj = v[cols[jj]]
                    for t in range(dd):
                        acc[r][t] += e * vj[t]
                new_l = (math.exp(lse_i[r] - m_ij) if m_ij != NEG_INF else 0.0) + l_ij
                m_i[r] = m_ij
                lse_i[r] = m_ij + math.log(new_l) if new_l > 0.0 else NEG_INF
        for r in range(len(rows)):
            o_scale = math.exp(m_i[r] - lse_i[r]) if lse_i[r] != NEG_INF else 0.0
            out.append([acc[r][t] * o_scale for t in range(dd)])
            lse.append(lse_i[r])
    return out, lse, scale


def naive_lse(q, k, causal=False, softmax_scale=None, bias=None):
    """对照用的 log-sum-exp：log Σ exp(score)。"""
    sq, sk, d = len(q), len(k), len(q[0])
    scale = softmax_scale_for(d, softmax_scale)
    res = []
    for i in range(sq):
        scores = []
        for j in range(sk):
            s = dot(q[i], k[j]) * scale
            if bias is not None:
                s += bias[j]
            if causal and j > i:
                continue
            scores.append(s)
        mx = max(scores)
        tot = 0.0
        for s in scores:
            tot += math.exp(s - mx)
        res.append(mx + math.log(tot))
    return res
